src/library_pthread.js

```diff
@@ -1368,12 +1368,31 @@ var LibraryPThread = {
     // Allocate a buffer, which will be copied by the C code.
     var stack = stackSave();
     // First passed parameter specifies the number of arguments to the function.
-    var args = stackAlloc(numCallArgs * 8);
+    // When BigInt support is enabled, we must handle types in a more complex
+    // way, detecting at runtime if a value is a BigInt or not (as we have no
+    // type info here). To do that, add a "prefix" before each value that
+    // indicates if it is a BigInt, which effectively doubles the number of
+    // values we serialize for proxying. TODO: pack this?
+    var serializedNumCallArgs = numCallArgs {{{ WASM_BIGINT ? "* 2" : "" }}};
+    var args = stackAlloc(serializedNumCallArgs * 8);
     var b = args >> 3;
     for (var i = 0; i < numCallArgs; i++) {
-      HEAPF64[b + i] = arguments[2 + i];
+      var arg = arguments[2 + i];
+#if WASM_BIGINT
+      if (typeof arg === 'bigint') {
+        // The prefix is non-zero to indicate a bigint.
+        HEAP64[b + 2*i] = BigInt(1);
+        HEAP64[b + 2*i + 1] = arg;
```

Comment:
Assigning a bigint to a slot in HEAP64 just works?

Reply:
Yes, `HEAP64` is a `BigInt64Array`. That's the only thing you can assign to it without erroring basically.